edge_server/chronos/devices/mock.py

```diff
@@ -0,0 +1,225 @@
+"""
+Mock device implementations for testing and development.
+"""
+
+import math
+import random
+import time
+from typing import Any, Dict
+
+from chronos.config import cfg
+from chronos.devices.hardware import SerialDevice
+from chronos.logging import root_logger as logger
+
+
+class MockSerialDevice(SerialDevice):
+    """
+    Mock implementation of SerialDevice.
+    Simulates device behavior without requiring actual hardware connections.
+    """
+
+    def __init__(
+        self, id: int, name: str = None, portname: str = None, baudrate: int = None
+    ):
+        """Initialize the mock serial device."""
+        # Initialize parent with same parameters
+        super().__init__(id, portname, baudrate, name)
+        self._state = False  # Initial state is OFF
+        logger.info(f"Initialized mock serial device {id}")
+
+    def _send_command(self, command: str) -> str:
+        """
+        Override the _send_command method to simulate responses without
+        actually connecting to hardware.
+        """
+        # Extract the command type and device ID from the command
+        parts = command.strip().split()
+        if len(parts) < 2:
+            return "error"
+
+        if parts[0] == "relay":
+            if parts[1] == "read":
+                # Return the current state
+                return "on" if self._state else "off"
+            elif parts[1] == "on":
+                # Set state to ON
+                self._state = True
+                logger.info(f"Set mock device {self._id} to ON")
+                return "ok"
+            elif parts[1] == "off":
+                # Set state to OFF
+                self._state = False
+                logger.info(f"Set mock device {self._id} to OFF")
+                return "ok"
+
+        return "error"
+
+
+class MockModbusDevice:
```

Comment:
maybe this can inherit from the hardware class... I'm not sure. could be helpful to reduce maintenance burden if it did.